[BLOG] Top 5 for an Azure Developer: Serverless (#458)
pulumi/pulumi-hugo: themes/default/content/blog/top-5-things-for-azure-devs-serverless/index.md

---
title: "Top 5 an Azure Developer Needs to Know: Serverless"
date: 2021-07-30
meta_desc: "Building and deploying serverless functions for Azure developers"
meta_image: azure-top-5.png
authors:
    - sophia-parafina
tags:
    - azure functions
    - serverless

---

> In the [previous article]({{< relref "/blog/top-5-things-for-azure-devs-vms" >}}), we did a deep dive into virtual machines, first looking at creating and deploying a virtual machine using the Azure Portal, then repeating the process using infrastructure as code. We further demonstrated how to automate provisioning as part of cloud engineering's  build and deploy process.

Reviewer: ```suggestion
In the [previous article]({{< relref "/blog/top-5-things-for-azure-devs-vms" >}}), we did a deep dive into virtual machines, first looking at creating and deploying a virtual machine using the Azure Portal, then repeating the process using infrastructure as code. We further demonstrated how to automate provisioning as part of cloud engineering's build and deploy processes.
```

> This article will explore the other end of the cloud infrastructure with serverless, which is an on-demand fully-managed cloud architecture.

Reviewer: ```suggestion
This article will explore the other end of the cloud infrastructure with serverless, which is an on-demand, fully-managed cloud architecture.
```

Azure Functions are ideal for specific use cases that include:

- when demand varies unpredictably, serverless can scale automatically
- when a service is used infrequently, serverless can reduce costs by only charging per use
- serverless can eliminate maintenance costs
- when you have client-heavy applications, where most of the business logic is in the client
- if you have high latency tasks like data processing

Serverless is not ideal if the application runs constantly, needs a quick response (otherwise known as the [cold start problem]({{< relref "/blog/aws-lambda-provisioned-concurrency-no-cold-starts">}})), or is a long-running process that can exceed the [timeout period](https://docs.microsoft.com/en-us/azure/azure-functions/functions-scale#timeout) for your execution plan.

With those caveats, let's jump into creating a serverless function using the Azure Portal to understand how to write a function and how the pieces work together.

## Our first serverless function

> For this example, we'll use [VS Code](https://code.visualstudio.com/) and the [Azure Functions extension](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-azurefunctions).

Reviewer: ```suggestion
For this example, we'll use [Visual Studio Code](https://code.visualstudio.com/) and the [Azure Functions extension](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-azurefunctions).
```

I think that it's preferred to refer to it this way; alternately it's probably "VSCode"?

- Name your function, e.g., myHttpTrigger.
- Set the Authorization level to Anonymous

![New Project](./image2.png)

When finished, you'll have a template.

```typescript
import { AzureFunction, Context, HttpRequest } from "@azure/functions"

const httpTrigger: AzureFunction = async function (context: Context, req: HttpRequest): Promise<void> {
    context.log('HTTP trigger function processed a request.');
    const name = (req.query.name || (req.body && req.body.name));
    const responseMessage = name
        ? "Hello, " + name + ". This HTTP triggered function executed successfully."
        : "This HTTP triggered function executed successfully. Pass a name in the query string or in the request body for a personalized response.";

    context.res = {
        // status: 200, /* Defaults to 200 */
        body: responseMessage
    };

};

export default httpTrigger;
```

**Step 2**: (Optional) If you have Azure Functions Core tools installed, you can run the function locally. Press F5 or click on `Run` > `Start debugging`.

The service will run locally and provide you with a URL:

```bash
Functions:

        myHttpTrigger: [GET,POST] http://localhost:7071/api/myHttpTrigger
```

> Navigate to the URL in a browser or use curl.

Reviewer: ```suggestion
Navigate to the URL in a browser or use `curl`.
```

```bash
$ curl http://localhost:7071/api/myHttpTrigger
This HTTP triggered function executed successfully. Pass a name in the query string or in the request body for a personalized response.
```

> Press Control-C in the terminal to stop the debugger.

Reviewer: ```suggestion
Press `Control-C` in the terminal to stop the debugger.
```

Maybe this would look better. Maybe not. Your call :)

**Step 3**: Publish the function to Azure.

Make sure you are signed into Azure and click on `Deploy function app…`

![Deploy function](./image3.png)

- Select `Create a new Function App in Azure` and enter a name, "exampleHttpTrigger"
- Select a runtime stack, `Node.js 14 LTS`
- Select a location for new resources, East US

In the terminal pane, click on the `Output` tab, and you'll see the URL for the function. You can call the function with curl with the name parameter.

```bash
$ curl https://myazurehttptrigger.azurewebsites.net/api/myhttptrigger?name=Pulumipus
Hello, Pulumipus. This HTTP triggered function executed successfully.
```

> Don't forget to delete the function in the Azure Portal.

Reviewer: ```suggestion
Don't forget to delete the function in the Azure Portal!
```

![Azure Portal](./image1.png)

## Deploying Azure Functions with code

> Creating and deploying an Azure Function is a bit more involved than using the Azure Function extension with VS Code. What goes on behind the scene after you write the function is a process that includes:

Reviewer: ```suggestion
Creating and deploying an Azure Function is a bit more involved than using the Azure Function extension with VS Code. What goes on behind the scenes after you write the function is a process that includes:
```

the billing plan, the connection string to Azure storage, the Functions version (3), the runtime (Node.js), the runtime Node version (14.0), and the SAS url for the function code, we can build and deploy our Azure Function.

    ```typescript
    const app = new web.WebApp("fa", {
        resourceGroupName: resourceGroup.name,
        serverFarmId: plan.id,
        kind: "functionapp",
        siteConfig: {
            appSettings: [
                { name: "AzureWebJobsStorage", value: storageConnectionString },
                { name: "FUNCTIONS_EXTENSION_VERSION", value: "~3" },
                { name: "FUNCTIONS_WORKER_RUNTIME", value: "node" },
                { name: "WEBSITE_NODE_DEFAULT_VERSION", value: "~14" },
                { name: "WEBSITE_RUN_FROM_PACKAGE", value: codeBlobUrl },
            ],
            http20Enabled: true,
            nodeVersion: "~14",
        },
    });

    export const endpoint = pulumi.interpolate`https://${app.defaultHostName}/api/HelloNode?name=Pulumi`;
    ```

At the very end of the script, we export the URL to the service for you to trigger.

## Why is it so complicated?

> Isn't infrastructure as code supposed to simplify infrastructure deployments? This seems more effort than using the VS Code extension or even the Azure CLI on the surface. However, it's important to note that this is *boiler-plate* code that you can reuse to deploy multiple functions with different runtimes and Consumption Plans. The key to using infrastructure code is that it is repeatable, customizable, and a CI/CD can deploy changes automatically.

Reviewer: ```suggestion
Isn't infrastructure as code supposed to simplify infrastructure deployments? This seems more effort than using the Visual Studio Code extension or even [the Azure CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli) on the surface. However, it's important to note that this is *boiler-plate* code that you can reuse to deploy multiple functions with different runtimes and Consumption Plans. The key to using infrastructure code is that it is repeatable, customizable, and a CI/CD pipeline can deploy changes automatically.
```
